### PR TITLE
Remove team when sole owner and remove projects

### LINKED ISSUE #6655: Cleanup teams and projects when an account is deleted
Users are [reporting](https://app.frontapp.com/open/cnv_igbm1ze) not being able to create projects after deleting (and recreating) their Gitpod account.

The New Project repo list will show "already taken" for projects left behind in deleted accounts 

This happens because projects under deleted accounts were not cleaned up, preventing re-use of the repositories in those orphan projects.

### To Reproduce

- Delete an account which has an existing/working Project.
- Try to create a new Project (under a different account) using the same repo.
- Observe "already taken" for that repo.

### Expected Behavior when deleting an account

- [x] Projects under the User account should be deleted
- [x] Teams, and the projects under them, exclusively held by the account should be deleted
- [x] Teams which also have a another owner, should not be deleted

### PR DESCRIPTION
**Pull request by laushinka**
## Description
1. Removes user projects
2. Removes teams when user is sole owner of the teams
3. Removes projects of those sole-owned teams

## Related Issue(s)

Fixes #6655

## How to test
The testing is quite extensive.
Test 1:
1. Create a team
2. Add a member to that team
3. Add a project to that team
4. Set one team as owner, one as member
5. Have the owner delete their **user** account. Expectation: team and project should be deleted.

Test 2:
1. Redo steps 1-4 above.
2. Have the member delete their **user** account. Expectation: team and project should not be deleted.

Test 3:
1. Redo steps 1-3 above.
2. Set both members as owners.
3. Have one owner delete their **user** account. Expectation: team and project should not be deleted.

Test 4:
1. Add project to the user.
2. Delete the user. Expectation: project should be deleted, i.e. that project is available to be added again.

## Release Notes

```release-note
When a user deletes their account, own projects will be deleted and made available to be added again.
When the user is the sole owner of a team, that team and its projects will be deleted and made available to be added again.
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#6655*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jankeromnes** [comment]
Super cool, many thanks! Taking a look now 👀

File: components/server/src/user/user-deletion-service.ts
Comment (jankeromnes) [inline components/server/src/user/user-deletion-service.ts:154]: Minor suggestion: Maybe it's worth waiting for all projects to be removed before deleting the teams? I believe the `forEach` will just trigger multiple parallel promises on-the-fly without `await`ing them. If we wanted to wait for completion on these calls, I believe this could be rewritten like so:

```suggestion
        for (const team of ownedTeams) {
            const teamProjects = await this.projectDb.findTeamProjects(team.id);
            await Promise.all(teamProjects.map(project => this.projectDb.markDeleted(project.id)));
        }
```

**Review by jankeromnes — COMMENTED** [review]
Thanks for implementing this clean-up! I think this is going to be really helpful. ✨

I've left a few thoughts and comments in-line. Please have a look, and decide whether to accept or decline my suggestions. 🙂

Also, I tried running through your test protocol, but faced a few problems:

> Test 1:
> 1. Create a team

I logged in with GitHub and created a new team, and copied the invite URL.

> 2. Add a member to that team

I copied the invite URL, then opened a private browser tab, logged in with GitLab (created a new Gitpod account, which I needed to unblock), then visited the invite URL to join as a member.

> 3. Add a project to that team

I (perhaps sneakily) decided to add the projects as the new GitLab member (and not the original GitHub team owner). Doing this, I ran into https://github.com/gitpod-io/gitpod/issues/6743 (probably unrelated to this PR, but maybe this is where things went wrong)

> 4. Set one team as owner, one as member

This was already the case (the original GitHub team creator was already the owner, and the new GitLab invited member was already a member) so no action needed.

> 5. Have the owner delete their **user** account. Expectation: team and project should be deleted.

When I deleted the original GitHub user account, I found in my private window (where the GitLab member was still logged in) that the team still existed, now without any owner, and that also the projects still existed:

<table><tr><td>
Screenshot 2021-11-17 at 11 19 49
</td><td>
Screenshot 2021-11-17 at 11 20 01
</td></tr></table>

Maybe I did something wrong? Or maybe it's the problem I noticed in the SQL query (see below)

Happy to discuss and/or clarify and/or help! 😇

**Comment by laushinka** [comment]
@jankeromnes Thank you for all the insights! I will go through them one by one 🤗

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (laushinka) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:116]: @jankeromnes Oh god this was in the other planet of my SQL comfort zone and took so long to get 😭 Thanks for taking it apart, which I was hoping for. 
I tested this in the DB with a team that has both an owner and a member and it managed to find it, but I just tried this `tm_2.role` addition and it also managed to find it. What I definitely will do is write more unit tests that cover more of these cases, and test more in the DB by checking off a list of edge cases.

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (laushinka) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:121]: Ok this is objectively WAY better.

File: components/server/src/user/user-deletion-service.ts
Comment (laushinka) [reply components/server/src/user/user-deletion-service.ts:154]: You're absolutely right 🙏🏽

**Comment by laushinka** [comment]
@jankeromnes I've applied the changes and have tested different cases manually as well, including the one you mentioned above. Ready for another review 🙏🏽

**Comment by jankeromnes** [comment]
Many thanks Laurie! 🙏

Re-flagging myself for review so that I don't forget about this(!)

**Review by jankeromnes — APPROVED** [review]
Perfection! 🎯 ✨ Works as advertised!

Many thanks for teaching Teams & Projects to clean up after them 🧹 🙏

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 23629167ff47fd2535253cd2a7b2e8589de0d7c4</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6721#pullrequestreview-810231079" title="Approved">jankeromnes</a>*

Associated issue: *#6655*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
